feat: Create upgrading tool (#988)
vaadin/docs: articles/upgrading/common/16-17.asciidoc

[discrete]
==== Move annotations to AppShellConfigurator

> Configuring the app with certain annotations is no supported elsewhere than in a class that implements `AppShellConfigurator`.

Reviewer: ```suggestion
The only place where configuring the app with certain annotations is supported is in a class that implements [interfacename]#AppShellConfigurator#.

```

> This applies for v15+ bootstrap mode (the default), but not for v14 legacy bootstrapping. Instead of nondeterministic behavior and logging an error, the build will fail when any of the following annotations are discovered elsewhere than in an `AppShellConfigurator` class:

Reviewer: ```suggestion
This applies for v15+ bootstrap mode (the default), but not for v14 legacy bootstrapping. 
Rather than showing nondeterministic behavior and logging an error, the build will fail when any of the following annotations occur outside an `AppShellConfigurator` class:

```

```
Meta.class, PWA.class, Inline.class, Viewport.class, BodySize.class, Push.class
```

[discrete]
==== Replace obsolete APIs

[discrete]
===== A new API for Binding Items to Components:
- `HasDataProvider` and `HasItems` are now replaced by new `HasListDataView`, `HasLazyDataView` and `HasDataView` interfaces in `Grid`, `Select` and `CheckBoxGroup`. It will be also gradually replaced in other components which have an items binding.

> - `setDataProvider()` is now deprecated and overloaded `setItems` methods are recommended to use.

Reviewer: ```suggestion
- [methodname]#setDataProvider()# is now deprecated and it is recommended to use overloaded [methodname]#setItems()# methods.

```

- `setItems` methods now have a return type instead of void.
- `HasItemsAndComponents` interface has been replaced by  `HasItemComponents` in order to support the Data View API in in-memory binding components.
- `HasHierarchicalDataProvider` no longer has `setItems` overloads for `Collection`, `Stream` and `Array`.

[discrete]
===== URL parameters template feature:
- `BeforeEvent` has a bunch of new methods for forwarding, rerouting and getting the parameters. Some methods are now deprecated or removed.
- `RouteRegistry`, `SessionRouteRegistry` interfaces are now supplemented with new methods and deprecate getters for route layouts.

> - `com.vaadin.flow.server.startup.RouteTarget` has been completely removed. This class was internal and should not have been used by anyone. In case you have been using it, please create an issue describing what you needed it for.

Reviewer: ```suggestion
- [classname]#com.vaadin.flow.server.startup.RouteTarget# has been completely removed. 
This class was internal and should not have been used. 
If you have been using it, please create an issue describing what you needed it for.

```
